File: backend/core/access/visibility.py
"""가시성 규칙 — 순수 함수.

adapters/db/chunk_search.py 의 _권한_WHERE 와 **같은 판정**을 내야 한다.
두 벌이 존재하는 이유: SQL 은 검색을 사전 필터링하고(존재를 감춘다),
이 함수는 이미 나온 결과를 재검증한다(spec 5.4). 재검증이 SQL 을 다시
부르면 그건 검증이 아니라 같은 코드를 두 번 믿는 것이다.

두 벌이 어긋나면 재검증이 무의미해지므로 tests/test_search_integration.py
가 실제 DB 로 둘을 대조한다.

Document 가 아니라 두 필드를 받는 이유: PolicyHit 도 Document 도 이 규칙을
쓰는데 둘은 다른 타입이다. 공통 조상을 만들면 한 번 쓰는 추상화가 된다.
"""

from core.types import Principal


def visible(required_clearance: int, allowed_departments: tuple[str, ...], p: Principal) -> bool:
    """이 주체에게 보이는가.

    허용부서가 비어 있으면 전사 공개다 — "아무도 못 본다"가 아니다.
    W1 실측: 빈 배열을 안 받아주면 그 문서는 자기 부서에도 안 보이고,
    에러가 아니라 조용한 누락이라 발견이 늦다.
    """
    return required_clearance <= p.clearance and (
        not allowed_departments or p.department in allowed_departments
    )

File: frontend/lib/visibility.ts
import type { Principal } from "../components/PersonaSegment.tsx";

// GET /documents 가 돌려주는 형태 그대로 — 권한 필터 없이 전부.
export type Document = {
  id: number;
  title: string;
  doc_type: string;
  required_clearance: number;
  allowed_departments: string[];
  source_path: string;
  chunk_count: number;
};

// backend/core/access/visibility.py 의 규칙과 같아야 한다 — SQL(chunk_search.
// _권한_WHERE) · 파이썬(visibility.py)에 이은 세 번째 사본. 허용 부서가
// 비어 있으면 전사 공개다("아무도 못 본다"가 아니다).
//
// `lib/` 에 있는 이유가 둘이다. 서버 컴포넌트가 불러야 하고(클라이언트
// 모듈에 두면 못 부른다), 세 번째 사본인데 여기 오기 전까지 테스트가
// 하나도 없었다 — npm test 는 lib/ 만 본다.
//
// `p` 가 없을 수 있다. 타입은 Principal 을 요구하지만 (explain)/documents 의
// `personas.find(...) ?? personas[0]` 은 목록이 비면 undefined 를 흘린다 —
// 실측: TypeError: Cannot read properties of undefined (reading 'clearance').
// 앱에 error.tsx 가 없어 그 예외가 화면 전체를 Next 기본 오류로 바꿨다.
// 여기서는 **닫히는 쪽**으로 답한다 — 주체를 모르면 보이지 않는다.
export function visible(d: Document, p: Principal | undefined): boolean {
  if (!p) return false;
  return (
    d.required_clearance <= p.clearance &&
    (d.allowed_departments.length === 0 || d.allowed_departments.includes(p.department))
  );
}
